### PR TITLE
Restore rectangle search

### LINKED ISSUE #2659: Re-enable rectangular search for SDSS query_region
#2477 Removed rectangular search from `query_region` in `SDSS` and converted it to a cone search. However, a drawback was not significantly discussed:

- The rectangular search can use a much larger "radius" i.e. field size. The limit for a cone search is 3 arcmin while the limit for a rectangular search can be as large as 12 arcmin (or more).

With the sparse fields we usually query for extragalactic transients, this does not usually get close to row-size limits. 

The changes in #2477 were made without Depreciation warnings as well, which I understand with the pre-release model.. however it would have been nice to at least preserve the old endpoint in  function such as `query_rectangle_region` or in a parameter such as `rectangular: bool = False` (where setting it to True uses the old `query_region`) so that backwards compatibility can be maintained.

Anyway this is a Feature Request to add this back in, as the old version still works (can be tested by pinning `astroquery` to e.g., `v0.4.6`, but making the change to the endpoint being HTTPS, a recent fix from #2654):

```Python
from astroquery import sdss
sdss.conf.skyserver_baseurl = sdss.conf.skyserver_baseurl.replace("http://","https://")
working_sdss_table = sdss.SDSS.query_region(<pars>,radius=<large radius multiple times over the MAX_CROSSID_RADIUS>) 
```

In fact in SDSS rectangular search there is only a suggestion, not a hard limit, on "radius" of 0.2 degrees, which seems to be not enforced by radius but by rows returned. See e.g.,: https://skyserver.sdss.org/dr17/SearchTools/rect

### PR DESCRIPTION
**Pull request by weaverba137**
This PR closes #2659. Rectangle searches are passed as a SQL query to `query_sql()`.

- [X] Fix any existing broken tests.
- [x] Enable tests with rectangular search.
- [x] Test calling `query_region()` with either `radius` or `width`; not neither, not both.
- [x] Decide what defaults should be: *e.g.* what *should* happen if neither `radius` nor `width` are specified?
- [x] Decide on a definition of "rectangle": `cos(dec)` or not.
- [x] Verify that documentation is clear.
- [x] Change log, etc.

Other notes:

* I rearranged some of the documentation sections, even in functions unrelated to this change, because they did not strictly follow numpydoc standards.
* But it turns out that breaks `async_to_sync`, so I'll undo that for the purposes of this PR.

### CODE REVIEW

**Comment by weaverba137** [comment]
The `async_to_sync` functionality appears to assume that `Returns` is the last section of the docstring, but this violates the numpydoc standard. I'll eventually get around to restoring the `Returns`, but it would be more robust for `async_to_sync` to find and replace the `Returns` section, rather than just assuming it is at the end.

**Comment by keflavich** [comment]
A fix to that incorrect assumption would be welcome!

**Comment by weaverba137** [comment]
@keflavich, good, but I think I will keep that as a separate issue for now.

**Comment by weaverba137** [comment]
@keflavich, I need some help on this one. I can no longer determine why `async_to_sync` is not producing compilable doc strings, even after I've restored the doc string sections as closely as possible to the state prior to this PR (as well as the last tag).

**Comment by keflavich** [comment]
I see https://readthedocs.org/projects/astroquery/builds/19731041/ is the problem.

I'm not sure, but my first guess is that you've changed something like
```
"""
this
"""
```
to
```
"""this
"""
```
which has broken the `dedent` code?

I can dig a bit more later if that's not the soln

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:212]: try 
```suggestion
        """
        Used to query a region around given coordinates. Either `radius` or
```

Comment (weaverba137) [reply astroquery/sdss/core.py]: That helped, but now it seems to be having problems with lines like:
```
Note that there is a server-side limit of 3 arcmin on `radius`.
```
which is the standard way to refer to arguments to a function.  Instead it is being interpreted with the `py:obj` role.

**Comment by keflavich** [comment]
>
> That helped, but now it seems to be having problems with lines like:
>
> Note that there is a server-side limit of 3 arcmin on `radius`.
>
> which is the standard way to refer to arguments to a function. Instead it
> is being interpreted with the py:obj role.
>

try double instead of single backticks?

> --
Adam

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py]: PS, [The Numpy Style Guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) uses single backticks to refer to parameters.

Comment (keflavich) [reply astroquery/sdss/core.py]: We must have (I think unintentionally?) assigned single-backticks to be objects?  I don't think this was an intentional choice.  See the opening note on https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html.

Comment (weaverba137) [reply astroquery/sdss/core.py]: Right, I was seeing other evidence of that. Probably not worth trying to solve in this PR, so I'll just use double-backticks for now.

Comment (keflavich) [reply astroquery/sdss/core.py]: If you can figure out where we set that, let me know - this is something that has always confused and frustrated me too.

Comment (weaverba137) [reply astroquery/sdss/core.py]: https://github.com/astropy/sphinx-astropy/blob/main/sphinx_astropy/conf/v1.py#L80

Comment (weaverba137) [reply astroquery/sdss/core.py]: Documentation is compiling successfully now, & I created astropy/sphinx-astropy#58 to ask about why `default_role` is set.

**Comment by weaverba137** [comment]
@keflavich, *et al*., I still need to do some updates to the remote data tests, but I think it is time to move this out of draft status and start thinking about the unchecked boxes above:

* What should be the default for `query_region()` if neither `radius` nor `width` are specified? Currently it raises an exception.
* What is the definition of "rectangle". Currently, I implemented an improved version of the original rectangle from the previous tag (now handles RA wrap-around!). It does not currently implement `cos(dec)` corrections.
* Is the documentation sufficiently clear?

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:327]:
```suggestion
            raise ValueError("Either radius or width must be specified, not both.")
```

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:325]:
```suggestion
            raise ValueError("Either radius or width must be specified.")
```

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:349]: I think this achieves the same thing more tersely:
```suggestion
            width = u.Quantity(width, u.degree).value
```

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:360]:
```suggestion
                height = u.Quantity(height, u.degree).value
```

File: astroquery/sdss/core.py
Comment (keflavich) [inline astroquery/sdss/core.py:381]:
```suggestion
            for target in coordinates:
```

**Review by keflavich — CHANGES_REQUESTED** [review]
> What should be the default for query_region() if neither radius nor width are specified? Currently it raises an exception.

This is the right behavior

> What is the definition of "rectangle". Currently, I implemented an improved version of the original rectangle from the previous tag (now handles RA wrap-around!). It does not currently implement cos(dec) corrections.

The current implementation is fine, but ...

> Is the documentation sufficiently clear?

Yes, but include a note that the rectangle coordinates represent (ra +/- width, dec +/- height) as opposed to (ra +/- width/cos(dec), dec +/- height)


Otherwise, looks good.  I put in some minor code readability suggestions - if they do not achieve the desired behavior, though, feel free to reject.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:327]: Done.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:325]: Done.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:381]: Done.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:349]: OK, that's a nice trick. Done. Although for the record, two possible types of exception are raised: `width = u.Quantity('foo', u.degree).value` raises `TypeError`, while `width = u.Quantity('2 things', u.degree).value` raises `ValueError`.

File: astroquery/sdss/core.py
Comment (weaverba137) [reply astroquery/sdss/core.py:360]: Done, but see comment above.

**Comment by weaverba137** [comment]
@keflavich, I've implemented your suggestions and also run remote tests to verify those are working.  The final step is updating the documentation.

**Comment by weaverba137** [comment]
@keflavich, *et al.*, I've added further details to the documentation and the change log.  For some reason Codecov is saying the total coverage decreased, even though I increased the coverage of sdss/core.py to 100%.  The largest decrease appears to be in cds/core.py, but I don't know why that is.

File: docs/sdss/sdss.rst
Comment (keflavich) [inline docs/sdss/sdss.rst:23]:
```suggestion
Note use of the keyword argument ``spectro``, which requires matches to have
```

File: docs/sdss/sdss.rst
Comment (keflavich) [inline docs/sdss/sdss.rst:64]:
```suggestion
Finally, though, the constructed rectangles *do* account for RA wrap-around,
```

**Review by keflavich — APPROVED** [review]
I like the clarification text about the shape of the selection.  I have two very minor suggestions but I think this is good to merge.  I'm not really worried about the coverage decrease because it can't be related, but if anyone else has ideas, voice 'em

File: docs/sdss/sdss.rst
Comment (weaverba137) [reply docs/sdss/sdss.rst:23]: Good suggestion, done.

File: docs/sdss/sdss.rst
Comment (weaverba137) [reply docs/sdss/sdss.rst:64]: Good suggestion, done.

**Comment by weaverba137** [comment]
OK, I've committed the last two suggestions.

**Comment by weaverba137** [comment]
Reminder: this is waiting for final review.

**Comment by keflavich** [comment]
@bsipocz I'm good with this, but I'm wary of hitting 'merge' without your eyes on it.

**Review by bsipocz — APPROVED** [review]
All looks good, thanks!

**Pull request merged**
